---
title: "Summary"
author: "Agatha Nabilla L"
date: "5/23/2022"
output:
  pdf_document:
    toc: yes
  html_notebook:
    toc: yes
  html_document:
    df_print: paged
    toc: yes
---

# Spatial Transcriptomics 

## Introduction

Spatial transcriptomics (ST) is a recently developed method to observe mRNA expression of the number of genes within the spatial context of tissues and cells. Introduced in 2016, it adopts the in-situ hybridization popularized in the 1960s, but to date, there are currently five categories known to obtain spatial information and only two commercially available technologies that support spatial transcriptomic analysis ; the 10X Visium and the Nanostring GeoMX [(Moses and Pachter, 2022; ](https://doi.org/10.1038/s41592-022-01494-3)[Li and Wang, 2021)](10.1038/s41368-021-00146-0). 

The first category is ROI (region of interest) selection or also familiar with micro-dissected gene expression. This technique allows isolation of ROI within the sample with UV laser cutting or fusion of tissue with a membrane by IR laser. Example of this method is TOMO-seq and IR LCM. The second is FISH (fluorescent in-situ hybridization). Also known as single-molecule FISH, the most popular method in this category is MERFISH, which uses a different barcoding strategy and is faster than seqFISH [(Moses and Pachter, 2022)](https://doi.org/10.1038/s41592-022-01494-3). However, the drawback of this method is that it does not provide the information of full transcriptome and needs specialized equipment, which makes it costly [(Asp, Bergenstråhle and Lundeberg, 2020)](https://doi.org10.1002/bies.201900221). The third is ISS (in-situ sequencing). This method performs RNA sequencing while the cells stay in the tissue, and the sequenced RNA is visualized with fluorescent probes. An example of this method is STARmap (Spatially resolved transcript amplicon read-out mapping), which uses the barcoded padlock probes that hybridize to the targets and add a second primer to avoid the reverse transcription from occurring. The limitation of this method is the limited field of view and the limited quantity of the target. Thus, it also does not support the full transcriptome analysis (Asp, Bergenstråhle, and Lundeberg, 2020)](https://doi.org10.1002/bies.201900221). The fourth is the de-novo reconstruction of spatial information, and the last is NGS with spatial barcoding [(Moses and Pachter, 2022)](https://doi.org/10.1038/s41592-022-01494-3). The de-novo reconstruction is a method that computationally pushes back the map of single cells into space with limited or no prior knowledge about the position of the sequenced cells. Although this method is compatible with the scRNA-seq, it is not compatible with human tissue because it needs a fusion construct to be recombinantly expressed in the cell of interest [(Fazal et al., 2019)](https://doi.org/10.1016/j.cell.2019.05.027). The NGS, with spatial barcoding or also familiar with in-situ capturing, is a method that preserves spatial locations of the target by capturing and barcoding the transcripts from tissue sections, followed by a sequencing process that is conducted on the outside of the tissue. The imaging process of the tissue then supports the overlaying of transcriptomics data and the tissue image. This process supports transcriptomics information in the targeted regions [(Asp, Bergenstråhle and Lundeberg, 2020)](https://doi.org10.1002/bies.201900221). 

Based on the spatial resolution, the spatially resolved transcriptomics is differentiated from the cellular, like the LCM, anatomical features like the TOMO-seq and 10X Visium, and sub-cellular, like MERFISH, APEX-seq, and STARmap, and tissue section limited to section thickness up to 10 µm like the Nanostring GeoMX. Based on the profiling approach,  10X-Visium differs from Nanostring GeoMX because it uses transcriptome-wide approach while GeoMX uses targeted approach  [(Asp, Bergenstråhle and Lundeberg, 2020)](https://doi.org10.1002/bies.201900221). 

## 10X Visium
Method and steps 
Generated data 
Process and analysis 

### Seurat
[here](/Users/agathanabillalestari/summer_research_project/trial_error_spaniel.Rmd)

### Spaniel
### Giotto

## Nanostring GeoMX
### GeoDiff 
### GeoMxWorkflows 
### GeomxTools 

# Trypanosoma brucei 

